src/vellum/workflows/monitoring/tests/workflow.py

```diff
@@ -0,0 +1,137 @@
+    subworkflow = IterationSubworkflow
+
+
+class SimpleMapExample(BaseWorkflow[Inputs, BaseState]):
+    graph = MapFruitsNode
+    _monitoring_initialized = False  # Class-level flag for the entire workflow system
+
+    class Outputs(BaseOutputs):
+        final_value = MapFruitsNode.Outputs.count
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        # Production-style: Top-level workflow initializes monitoring for all components
+        if not SimpleMapExample._monitoring_initialized:
+            self._initialize_monitoring_dynamic()
+            SimpleMapExample._monitoring_initialized = True
+
+    def _initialize_monitoring_dynamic(self):
+        """Dynamically discover and wrap all execution methods with proper type classification. -- TESTING ONLY."""
+
+        # Classes to monitor with their proper types
+        workflow_classes = [SimpleMapExample, IterationSubworkflow]
+        node_classes = [MapFruitsNode, Iteration]
+
+        wrapped_methods = set()
+
+        # Wrap workflow classes
+        for workflow_class in workflow_classes:
+            for method_name in self._get_execution_methods(workflow_class):
+                if hasattr(workflow_class, method_name):
+                    method = getattr(workflow_class, method_name)
+                    if callable(method) and not hasattr(method, "__wrapped__"):
+                        monitor_name = f"{workflow_class.__name__}.{method_name}"
+
+                        # Create a monitoring decorator with proper type classification
+                        decorator = self._create_workflow_monitor(monitor_name)
+                        wrapped_method = decorator(method)
+                        setattr(workflow_class, method_name, wrapped_method)
+                        wrapped_methods.add(monitor_name)
+
+        # Wrap node classes
+        for node_class in node_classes:
+            for method_name in self._get_execution_methods(node_class):
+                if hasattr(node_class, method_name):
+                    method = getattr(node_class, method_name)
+                    if callable(method) and not hasattr(method, "__wrapped__"):
+                        monitor_name = f"{node_class.__name__}.{method_name}"
+
+                        # Create a monitoring decorator with proper type classification
+                        decorator = self._create_node_monitor(monitor_name)
+                        wrapped_method = decorator(method)
+                        setattr(node_class, method_name, wrapped_method)
+                        wrapped_methods.add(monitor_name)
+
+        # Successfully wrapped all discovered execution methods
+
+    def _get_execution_methods(self, cls: Type[Any]) -> List[str]:
+        """Get list of methods that should be monitored for a class."""
+        execution_methods = []
+
+        # Standard execution methods
+        standard_methods = ["run", "stream", "_run_subworkflow", "_context_run_subworkflow"]
```

Comment:
Thinking we may want to define these on `BaseNode` and `BaseWorkflow` instead of hardcoding within the helper.  We should also pick one subworkflow-running method to wrap and standardize our nodes around it